enhancements/per-node-configuration.md

```diff
@@ -0,0 +1,656 @@
+3. **Handle validation failures**:
+   - Log validation errors
+   - Set appropriate status conditions on the failing resource(s)
+   - Skip configuration application for invalid resources
+   - Requeue for retry (in case it's a transient race condition)
+
+### Scalability Considerations
+
+- **Node Watch Overhead**: The controllers watch Node resources for label changes. In clusters with frequent node label updates, this may increase reconciliation frequency. We can improve this by making each daemonset pod reconcile only on the running node using controller-runtime cache filter mechanism.
+- **Selector Evaluation**: For each reconciliation, the controller evaluates node selectors against the node where the daemonset is running, so reconciliation is done in parallel between the nodes.
+
+## Implementation Plan
+
+**API Changes and Validation:**
+- Add `NodeSelector` field to all CRD Spec types:
+  - `UnderlaySpec` (api/v1alpha1/underlay_types.go)
+  - `L3VNISpec` (api/v1alpha1/l3vni_types.go)
+  - `L2VNISpec` (api/v1alpha1/l2vni_types.go)
+  - `L3PassthroughSpec` (api/v1alpha1/l3passthrough_types.go)
+- Update CRD generation (`make manifests`)
+- Implement validation webhook for Underlay overlapping selector detection (only)
+- Add unit tests for validation logic
+
+**Controller Implementation:**
+- Add Node watch to all controllers (Underlay, L3VNI, L2VNI, L3Passthrough), caching only the node where the daemonset pod is running
+- Implement node matching logic using `metav1.LabelSelector` in each controller
+- **Implement reconcile loop validation** (following the existing pattern in `internal/controller/routerconfiguration/underlay_vni_controller.go`):
+  - Add validation calls in the reconcile loop for all CRD types
+  - For Underlay: validate no overlapping node selectors
+  - For L3VNI, L2VNI, L3Passthrough: validate other business rules (VNI conflicts, host session conflicts, etc.)
+  - On validation failure: log errors, set status conditions, skip configuration, return without error to avoid infinite retry
+  - Ensure validation uses the same logic as the admission webhook
+- Update configuration generation to be node-aware for each controller:
+  - Underlay: FRR BGP configuration
+  - L3VNI: VXLAN and VRF configuration
+  - L2VNI: Bridge and VXLAN configuration
+  - L3Passthrough: Host session configuration
+- Handle dynamic node label changes in all controllers
+
+**Documentation and Examples:**
+- Update API documentation for all CRDs
+- Create example configurations for:
+  - **Underlay**: Multi-rack, multi-datacenter, mixed hardware deployments
+  - **L3VNI**: Per-rack VNI, zone-specific VNI configurations
+  - **L2VNI**: Selective deployment, per-rack gateway configurations
+  - **L3Passthrough**: Edge nodes, per-security-zone configurations
+- Migration guide from cluster-wide to per-node configuration
+- Troubleshooting guide for selector conflicts
```

Comment:
I think we can offload this discussion to the PR.

Reply:
I will remove the whole section